File: cs/2/getpasswd.cs
using System;
using System.IO;
using System.Text;
using System.Linq;
using System.Collections.Generic;

namespace aic
{
    class GetPasswd
    {        
        static void Main(string[] args)
        {
            if (args.Length<1)
            {
                Console.WriteLine($"Usage: getpasswd input.txt");
                System.Environment.Exit(1);
            }
            var filename = args[0];
            Console.WriteLine($"Opening File: {filename}");
            if (!File.Exists(filename))
            {
                Console.WriteLine($"File {filename} does not exist");
                System.Environment.Exit(1);
            }
            var fileStream = new FileStream(args[0], FileMode.Open, FileAccess.Read);
            var result = 0;
            using (var streamReader = new StreamReader(fileStream, Encoding.UTF8))
            {
                string line;
                while ((line = streamReader.ReadLine()) != null) 
                {
                    // First item is rule, second is password
                    var items = line.Trim().Split(':');
                    var password = items[1].Trim();
                    var rules = items[0].Split(' ');
                    var letter = rules[1].Trim();
                    var count = rules[0].Split('-');
                    var min = Int16.Parse(count[0]);
                    var max = Int16.Parse(count[1]);
                    Console.WriteLine($"Password: {password} must be {min}-{max} {letter}'s in it");
                    // Now, count the number of "letters" found
                    var lettercount = 0;
                    for (var i=0; i<password.Length; i++)
                    {
                        if (password[i] == letter[0])
                        {
                            lettercount++;
                        }
                    }
                    Console.WriteLine($"Got {lettercount} items");
                    if (lettercount>=min && lettercount<=max)
                    {
                        Console.WriteLine("Got one!");
                        result++;
                    }
                }
            }
            Console.WriteLine($"There are {result} good passwords");
        }
    }
}

File: js/2/getpasswd.js
//
// Adventure in Code task 2
//

const fs = require('fs');
const yargs = require('yargs');
const format = require('string-format')

var myArgs = process.argv.slice(2);

const argv = yargs
    .option('filename', {
        alias: 'f',
        description: 'Filename to load',
        type: 'string',
        demandOption: true
    })
    .help()
    .alias('help', 'h')
    .argv;

if (argv.filename)
{
    console.log("Using filename: " + argv.filename);
}

var array = fs.readFileSync(argv.filename).toString().split("\n");
var count = 0;
for(i in array)
{
    if (array[i] == '')
    {
        continue;
    }
    var tokens = array[i].split(" ");
    // First item is the min/max
    var minmax = tokens[0].split("-");
    var min = parseInt(minmax[0], 10);
    var max = parseInt(minmax[1], 10);
    // Second item is the letter
    var letter = tokens[1].substr(0, 1);
    var passwd = tokens[2];
    var lettercount = 0;
    for (i in passwd)
    {
        if (passwd[i] == letter)
        {
            lettercount++;
        }
    }
    if ((lettercount>=min) && (lettercount<=max))
    {
        count++;
    }
}
console.log(format('Found {} valid passwords', count));
